rust/src/mqtt/mqtt.rs

```diff
@@ -0,0 +1,846 @@
+
+        while current.len() > 0 {
+            let mut skipped = false;
+            SCLogDebug!("response: handling {}", current.len());
+            match parse_message(current, self.protocol_version, self.max_msg_len as usize) {
+                Ok((rem, msg)) => {
+                    SCLogDebug!("response msg {:?}", msg);
+                    match msg.op {
+                        MQTTOperation::TRUNCATED(ref trunc) => {
+                            SCLogDebug!("found truncated with skipped {} current len {}", trunc.skipped_length, current.len());
+                            if trunc.skipped_length >= current.len() {
+                                skipped = true;
+                                self.skip_response = trunc.skipped_length - current.len();
+                            } else {
+                                current = &current[trunc.skipped_length..];
+                                self.skip_response = 0;
+                            }
+                            SCLogDebug!("skip_response now {}", self.skip_response);
+                        },
+                        _ => {}
+                    }
+                    self.handle_msg(msg, true);
+                    if skipped {
+                        return AppLayerResult::ok();
+                    }
+                    consumed += current.len() - rem.len();
+                    current = rem;
+                }
+                Err(nom::Err::Incomplete(_)) => {
+                    SCLogDebug!("incomplete response: consumed {} needed {} (input len {})", consumed, (current.len() + 1), input.len());
+                    return AppLayerResult::incomplete(consumed as u32, (current.len() + 1) as u32);
+                }
+                Err(_) => {
+                    return AppLayerResult::err();
+                }
+            }
+        }
+
+        return AppLayerResult::ok();
+    }
+
+    fn set_event(tx: &mut MQTTTransaction, event: MQTTEvent) {
+        let ev = event as u8;
+        core::sc_app_layer_decoder_events_set_event_raw(&mut tx.events, ev);
+    }
+
+    fn tx_iterator(
+        &mut self,
+        min_tx_id: u64,
+        state: &mut u64,
+    ) -> Option<(&MQTTTransaction, u64, bool)> {
+        let mut index = *state as usize;
+        let len = self.transactions.len();
+
+        while index < len {
+            let tx = &self.transactions[index];
+            if tx.tx_id < min_tx_id + 1 {
+                index += 1;
+                continue;
+            }
+            *state = index as u64;
+            return Some((tx, tx.tx_id - 1, (len - index) > 1));
+        }
+
+        return None;
+    }
+}
+
+// C exports.
+
+export_tx_get_detect_state!(rs_mqtt_tx_get_detect_state, MQTTTransaction);
+export_tx_set_detect_state!(rs_mqtt_tx_set_detect_state, MQTTTransaction);
+
+#[no_mangle]
+pub extern "C" fn rs_mqtt_probing_parser(
+    _flow: *const Flow,
+    _direction: u8,
+    input: *const u8,
+    input_len: u32,
+    _rdir: *mut u8,
```

Comment:
Can we know the direction of a mqtt message ? and so set `rdir` accordingly 
cf https://github.com/OISF/suricata/pull/5093 and more precisely https://github.com/OISF/suricata/pull/5093/commits/182ae51f483e5e7dca4f0fb83b97b96a3a00ccb0

Reply:
I didn't know about `rdir` before and will take a look. In general, the direction is defined by the initial communication (`CONNECT`) which identifies the broker and client. Afterwards, both sides can exchange identical messages between themselves (e.g. both can `PUBLISH`).

Reply:
Ok, so I am not sure it makes sense here...